janitor/utils.py

```diff
@@ -1680,3 +1687,176 @@ def _computations_as_categorical(df: pd.DataFrame, **kwargs) -> pd.DataFrame:
+    if not start_check:
+        raise ValueError(
+            """
+            The start value for the slice
+            must either be a string or `None`.
+            """
+        )
+    if not stop_check:
+        raise ValueError(
+            """
+            The stop value for the slice
+            must either be a string or `None`.
+            """
+        )
+    if not step_check:
+        raise ValueError(
+            """
+            The step value for the slice
+            must either be an integer or `None`.
+            """
+        )
+    start_check = any((start is None, start in df.columns))
+    stop_check = any((stop is None, stop in df.columns))
+    if not start_check:
+        raise ValueError(
+            """
+            The start value for the slice must either be `None`
+            or exist in the dataframe's columns.
+            """
+        )
+    if not stop_check:
+        raise ValueError(
+            """
+            The stop value for the slice must either be `None`
+            or exist in the dataframe's columns.
+            """
+        )
+    filtered_columns = df.columns.slice_locs(start=start, end=stop)
+    # slice_locs fails when step has a value
+    # so this extra step is necessary to get the correct output
+    start, stop = filtered_columns
+    filtered_columns = df.columns[slice(start, stop, step)]
+    return filtered_columns
+
+
+@_select_columns.register(dispatch_callable)
+def _(columns_to_select, df):
+    """
+    Base function for column selection.
+    Applies only to callables.
+    The callable is applied to every column in the dataframe.
+    Either True or False is expected per column.
+    A list/pd.Index of column names is returned.
+    """
+    # the function will be applied per series.
+    # this allows filtration based on the contents of the series
+    # or based on the name of the series,
+    # which happens to be a column name as well.
+    # whatever the case may be,
+    # the returned values should be a sequence of booleans,
+    # with at least one True.
+    filtered_columns = None
+    filtered_columns = [columns_to_select(column) for _, column in df.items()]
+
+    # returns numpy bool,
+    # which does not work the same way as python's bool
+    # as such, cant use isinstance.
+    # pandas type check function helps out
+    checks = (pd.api.types.is_bool(column) for column in filtered_columns)
+
+    if not all(checks):
+        raise ValueError(
+            "The callable provided must return a sequence of booleans."
+        )
+
+    # cant use `is` here, since it may be a numpy bool
+    checks = any((column == True for column in filtered_columns))  # noqa: E712
+
+    if not checks:
+        raise ValueError("No results were returned for the callable provided.")
+
+    filtered_columns = df.columns[filtered_columns]
+    return filtered_columns
+
+
+# hack to get it to recognize typing.Pattern
```

Comment:
That is quite the hack. In the interest of maintainability, would you be up for explaining the `re.compile` piece? I'm not well-versed enough in regex to know what's happening - what is being searched for? How come the "type of a re.compile" works here? 

If there's a reference you used that explains this piece, can you stick it in the docstring too?

Reply:
type(re.compile(r"\d")) returns the type for a regex, which is re.Pattern. functools.singledispatch has no problem with that; if however, I typed in re.Pattern, it associates it instead with the typing module and fails.